controllers/state_of_the_world.go

```diff
@@ -0,0 +1,149 @@
+		controller.WithRunnable("httproute watcher", controller.Watch(&gwapiv1.HTTPRoute{}, controller.HTTPRoutesResource, metav1.NamespaceAll)),
+		controller.WithRunnable("dnspolicy watcher", controller.Watch(&kuadrantv1alpha1.DNSPolicy{}, kuadrantv1alpha1.DNSPoliciesResource, metav1.NamespaceAll)),
+		controller.WithRunnable("tlspolicy watcher", controller.Watch(&kuadrantv1alpha1.TLSPolicy{}, kuadrantv1alpha1.TLSPoliciesResource, metav1.NamespaceAll)),
+		controller.WithRunnable("authpolicy watcher", controller.Watch(&kuadrantv1beta2.AuthPolicy{}, kuadrantv1beta2.AuthPoliciesResource, metav1.NamespaceAll)),
+		controller.WithRunnable("ratelimitpolicy watcher", controller.Watch(&kuadrantv1beta2.RateLimitPolicy{}, kuadrantv1beta2.RateLimitPoliciesResource, metav1.NamespaceAll)),
+		controller.WithRunnable("configmap watcher", controller.Watch(&corev1.ConfigMap{}, controller.ConfigMapsResource, kuadrantNamespace, controller.FilterResourcesByLabel[*corev1.ConfigMap]("kuadrant.io/topology=true"))),
+		controller.WithPolicyKinds(
+			kuadrantv1alpha1.DNSPolicyKind,
+			kuadrantv1alpha1.TLSPolicyKind,
+			kuadrantv1beta2.AuthPolicyKind,
+			kuadrantv1beta2.RateLimitPolicyKind,
+		),
+		controller.WithObjectKinds(
+			kuadrantv1beta1.KuadrantKind,
+			schema.GroupKind{Group: corev1.GroupName, Kind: "ConfigMap"}),
+		controller.WithObjectLinks(
+			kuadrantv1beta1.LinkKuadrantToGatewayClasses,
+			kuadrantv1beta1.LinkKuadrantToTopologyConfigMap,
+		),
+		controller.WithReconcile(buildReconciler(client)),
+	}
+
+	return controller.NewController(controllerOpts...)
+}
+
+func buildReconciler(client *dynamic.DynamicClient) controller.ReconcileFunc {
+	reconciler := &controller.Workflow{
+		Precondition: NewEventLogger().Log,
+		Tasks: []controller.ReconcileFunc{
+			NewTopologyFileReconciler(client).Reconcile,
+		},
+	}
+
+	return reconciler.Run
+}
+
+type TopologyFileReconciler struct {
+	Client *dynamic.DynamicClient
+}
+
+func NewTopologyFileReconciler(client *dynamic.DynamicClient) *TopologyFileReconciler {
+	return &TopologyFileReconciler{Client: client}
+}
+
+func (r *TopologyFileReconciler) Reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology) {
+	logger := controller.LoggerFromContext(ctx).WithName("topology file")
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "topology",
+			Namespace: kuadrantNamespace,
+			Labels:    map[string]string{"kuadrant.io/topology": "true"},
+		},
+		Data: map[string]string{
+			"topology": topology.ToDot(),
+		},
+	}
+
+	configMapRes := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "configmaps"}
+	unstructuredCM, _ := controller.Destruct(cm)
+
+	targets := topology.Objects().Items(func(object machinery.Object) bool {
+		if object.GetName() == cm.GetName() && object.GetNamespace() == cm.GetNamespace() {
+			return true
+		}
+		return false
```

Comment:
You probably want to filter by kind too:

```suggestion
		_, ok := o.(*corev1.ConfigMap)
		return ok && object.GetName() == cm.GetName() && object.GetNamespace() == cm.GetNamespace()

```

Reply:
Or yet the following.

Say you name the desired topology ConfiMap `topologyConfigMap` instead of `cm`

```go
topologyConfigMap := &corev1.ConfigMap{
	ObjectMeta: metav1.ObjectMeta{
		Name:      "topology",
		Namespace: kuadrantNamespace,
		Labels:    map[string]string{"kuadrant.io/topology": "true"},
	},
	Data: map[string]string{
		"topology": topology.ToDot(),
	},
}

[…]

topologyConfigMaps := lo.FilterMap(objs, func(obj machinery.Object, _ int) (*corev1.ConfigMap, bool) {
	cm, ok := obj.(*corev1.ConfigMap)
	return cm, ok && cm.GetNamespace() == topologyConfigMap.GetNamespace() && cm.GetName() == topologyConfigMap.GetName()
})

if len(topologyConfigMaps) == 0 {
	// …
}
…
```

I guess this is the option I like the most. WDYT?

Reply:
To be even more accurate in the search if you want, given the topology ConfigMap must be linked to the Kuadrant object, first find the Kuadrant object in the topology, then...

```go
existingTopologyConfigMaps := lo.FilterMap(topology.Objects().Children(kuadrantObj), …)
```

Reply:
Finally, just a friendly reminder that the Kuadrant object is a root object in the topology. So...

```go
kuadrantObject := topology.Objects().Root()[0]
```

🙂

Reply:
A couple of things here. Firstly the refactor to have to remove the if block and have a direct return is of course something that I can do.

For the rest of the suggestions I have issue with them.

The checking if the object is in fact a kind of configmap makes sense but when I try the type assertion I am getting the following error: `Impossible type assertion: '*corev1.ConfigMap' does not implement 'machinery.Object'`. It is quite possible I am doing something wrong but I would assume `_, ok := object.(*corev1.ConfigMap)` would be the correct way.

If the preferred filter method is use `lo.FitlerMap(topology.Objects().Items(), func()....` do we have a design flaw in the policy-machinery by allowing the user to pass in a filter function into `topology.Objects().Items()`?

The last point is on `kuadrantObject := topology.Objects().Root()[0]`, this may not be true. The topology allows for multiple kuadrant CRs. While currently today the operator only support one instance of the Kuadrant CR, I am not sure how long this will hold true. I know there has being talks about supporting multiple gateways by using the different Kuadrant CRs for that. I am well aware that this will not be happening for a long time, but we would be wrong to make the assumption that there will only ever be one Kuadrant CR.

Reply:
> The checking if the object is in fact a kind of configmap makes sense but when I try the type assertion I am getting the following error: `Impossible type assertion: '*corev1.ConfigMap' does not implement 'machinery.Object'`. It is quite possible I am doing something wrong but I would assume `_, ok := object.(*corev1.ConfigMap)` would be the correct way.

Actually, the `Impossible type assertion` error makes sense. This is because the object is actually a [`controller.RuntimeObject`](https://github.com/Kuadrant/policy-machinery/blob/85837a3f6a5ae4b84594de2124c5338f678a57f3/controller/object.go#L17) which wraps a `*corev1.ConfigMap`. It cannot be a `*corev1.ConfigMap` indeed. So my previous suggestion was misleading. Instead, you'd need something like this:

```go
ro, ok := obj.(*RuntimeObject)
if !ok
	return nil, false
end
cm, ok := ro.Object.(*corev1.ConfigMap)
return cm, ok && cm.GetNamespace() == topologyConfigMap.GetNamespace() && cm.GetName() == topologyConfigMap.GetName()
```

Another option is checking the `GroupKind`, tho you'd still need to make the conversion to a ConfigMap in a separate step:

```go
obj.GroupVersionKind().GroupKind() == ConfigMapGroupKind
```

<br/>

> If the preferred filter method is use `lo.FitlerMap(topology.Objects().Items(), func()....` do we have a design flaw in the policy-machinery by allowing the user to pass in a filter function into `topology.Objects().Items()`?

I don't see it as a design flaw. If all you want is filtering, then you can pass a predicate to the `Items()` func. I proposed using `lo.FilterMap` instead because in this case not only we want to filter, but also we want the objects converted to `*corev1.ConfigMap` if possible.

That said, if this is something we believe we'll be doing a lot, it may be a good addition to the Policy Machinery.

<br/>

> The last point is on `kuadrantObject := topology.Objects().Root()[0]`, this may not be true. The topology allows for multiple kuadrant CRs. While currently today the operator only support one instance of the Kuadrant CR, I am not sure how long this will hold true. I know there has being talks about supporting multiple gateways by using the different Kuadrant CRs for that. I am well aware that this will not be happening for a long time, but we would be wrong to make the assumption that there will only ever be one Kuadrant CR.

This is the hardest one to unpack IMO.

First, we need to decide if we want to allow more than one Kuadrant object per cluster or not and what to do in case of not allowing it.

Allowing multiple Kuadrant _objects_ does not necessarily mean allowing multiple Kuadrant _instances_. We can let users create as many Kuadrant CRs they want, but commit to reconcile only the first that gets created, leaving all the others stale (maybe just updating their status). These objects would be loose roots to which nothing would be linked.

Another option is not to allow multiple Kuadrant objects. Be harsh by perhaps making the reconciliation workflow to force delete any second CR of this kind and on, keeping only one present at a time. Not very user-friendly, but a possibility nonetheless.

Either way, the reason to structure the topology as a DAG is so we can navigate it from the root, "making assumptions" about the structure.

If we have to list and filter all resources then not only this could be computationally costly, but also we might very well have simply listed the resources in a map.

I guess what I'm trying to say is that we want to read resources from the topology by navigating the DAG, not in the traditional way that the API server would otherwise let us. Each time we use a func such as `Parent` or `Children`, there's a semantic embedded into it, whose counterpart is the linking functions. We want to use those functions and thus ensure the objects handled in the process are where they are supposed to be.

Reply:
we have already some code that assumes there is only one kuadrant CR. If that simplifies your work, feel free to assume it. Drop a comment and we can easily spot that later if the assumption happens be no longer true in the future.

Reply:
I agree with @eguzki. Let users create as many Kuadrant CRs as they want, have a well-commented func that filters amongst all roots of the topology and returns the oldest of the Kuadrant CRs only, use this func to link the GatewayClasses to this one Kuadrant CR only and ignore all the others, use the func to retrieve the one and only root of the DAG whenever we want to navigate it (to read other resources).